diff -urNp -x '*.orig' sssd-1.13.4.org/src/providers/ad/ad_gpo_ndr.c sssd-1.13.4/src/providers/ad/ad_gpo_ndr.c
--- sssd-1.13.4.org/src/providers/ad/ad_gpo_ndr.c	2016-04-13 16:48:41.000000000 +0200
+++ sssd-1.13.4/src/providers/ad/ad_gpo_ndr.c	2021-03-03 21:59:14.896225612 +0100
@@ -258,7 +258,7 @@ ndr_pull_dom_sid(struct ndr_pull *ndr,
         NDR_CHECK(ndr_pull_align(ndr, 4));
         NDR_CHECK(ndr_pull_uint8(ndr, NDR_SCALARS, &r->sid_rev_num));
         NDR_CHECK(ndr_pull_int8(ndr, NDR_SCALARS, &r->num_auths));
-        if (r->num_auths < 0 || r->num_auths > ARRAY_SIZE(r->sub_auths)) {
+        if (r->num_auths < 0 || r->num_auths > N_ELEMENTS(r->sub_auths)) {
             return ndr_pull_error(ndr, NDR_ERR_RANGE, "value out of range");
         }
         NDR_CHECK(ndr_pull_array_uint8(ndr, NDR_SCALARS, r->id_auth, 6));
diff -urNp -x '*.orig' sssd-1.13.4.org/src/util/util.h sssd-1.13.4/src/util/util.h
--- sssd-1.13.4.org/src/util/util.h	2016-04-13 16:48:41.000000000 +0200
+++ sssd-1.13.4/src/util/util.h	2021-03-03 21:59:14.896225612 +0100
@@ -587,4 +587,8 @@ int sss_unique_file(TALLOC_CTX *owner,
  */
 int sss_unique_filename(TALLOC_CTX *owner, char *path_tmpl);
 
+#ifndef N_ELEMENTS
+#define N_ELEMENTS(arr) (sizeof(arr) / sizeof(arr[0]))
+#endif
+
 #endif /* __SSSD_UTIL_H__ */
