diff -urNp -x '*.orig' sssd-1.13.4.org/src/lib/sifp/sss_sifp_utils.c sssd-1.13.4/src/lib/sifp/sss_sifp_utils.c
--- sssd-1.13.4.org/src/lib/sifp/sss_sifp_utils.c	2016-04-13 16:48:41.000000000 +0200
+++ sssd-1.13.4/src/lib/sifp/sss_sifp_utils.c	2021-03-03 21:59:14.489430078 +0100
@@ -41,7 +41,7 @@ void sss_sifp_set_io_error(sss_sifp_ctx
 {
     dbus_error_free(ctx->io_error);
     dbus_error_init(ctx->io_error);
-    dbus_set_error(ctx->io_error, error->name, error->message);
+    dbus_set_error(ctx->io_error, error->name, "%s", error->message);
 }
 
 const char *
