diff -ur sssd-1.6.1-o/src/external/python.m4 sssd-1.6.1/src/external/python.m4
--- sssd-1.6.1-o/src/external/python.m4	2011-08-29 09:39:05.000000000 -0600
+++ sssd-1.6.1/src/external/python.m4	2011-10-16 00:03:21.000000000 -0600
@@ -12,14 +12,8 @@
     AC_PATH_PROG(PYTHON, python)
     AC_MSG_CHECKING([for working python])
     if test -x "$PYTHON"; then
-        PYTHON_CFLAGS="`$PYTHON -c \"from distutils import sysconfig; \
-            print '-I' + sysconfig.get_python_inc() + \
-            ' -I' + sysconfig.get_python_inc(plat_specific=True) + ' ' + \
-            sysconfig.get_config_var('BASECFLAGS')\"`"
-        PYTHON_LIBS="`$PYTHON -c \"from distutils import sysconfig; \
-            print \\\" \\\".join(sysconfig.get_config_var('LIBS').split() + \
-            sysconfig.get_config_var('SYSLIBS').split()) + \
-            ' -lpython' + sysconfig.get_config_var('VERSION')\"`"
+        PYTHON_CFLAGS="`python-config --cflags`"
+        PYTHON_LIBS="`python-config --libs`"
             AC_MSG_RESULT([yes])
     else
         AC_MSG_ERROR([no. Please install python devel package])
Only in sssd-1.6.1/src/external: python.m4~
