diff -urNp -x '*.orig' sssd-1.13.4.org/Makefile.am sssd-1.13.4/Makefile.am
--- sssd-1.13.4.org/Makefile.am	2016-04-13 16:48:41.000000000 +0200
+++ sssd-1.13.4/Makefile.am	2021-03-03 21:59:12.948942197 +0100
@@ -3597,11 +3597,13 @@ if BUILD_PYTHON2_BINDINGS
 		cd $(builddir)/src/config; \
 		$(PYTHON2) setup.py build --build-base $(abs_builddir)/src/config \
 			install $(DISTSETUPOPTS) --prefix=$(PYTHON2_PREFIX) \
+			--install-purelib=$(python2dir) --optimize=2 \
 			--record=$(abs_builddir)/src/config/.files2; \
 	else \
 		cd $(builddir)/src/config; \
 		$(PYTHON2) setup.py build --build-base $(abs_builddir)/src/config \
 			install $(DISTSETUPOPTS) --prefix=$(PYTHON2_PREFIX) \
+			--install-purelib=$(python2dir) --optimize=2 \
 			--record=$(abs_builddir)/src/config/.files2 --root=$(DESTDIR); \
 	fi
 	cd $(DESTDIR)$(py2execdir) && \
@@ -3615,11 +3617,13 @@ if BUILD_PYTHON3_BINDINGS
 		cd $(builddir)/src/config; \
 		$(PYTHON3) setup.py build --build-base $(abs_builddir)/src/config \
 			install $(DISTSETUPOPTS) --prefix=$(PYTHON3_PREFIX) \
+			--install-purelib=$(python3dir) --optimize=2 \
 			--record=$(abs_builddir)/src/config/.files3; \
 	else \
 		cd $(builddir)/src/config; \
 		$(PYTHON3) setup.py build --build-base $(abs_builddir)/src/config \
 			install $(DISTSETUPOPTS) --prefix=$(PYTHON3_PREFIX) \
+			--install-purelib=$(python3dir) --optimize=2 \
 			--record=$(abs_builddir)/src/config/.files3 --root=$(DESTDIR); \
 	fi
 	cd $(DESTDIR)$(py3execdir) && \
