--- sssd-1.11.4/src/external/python.m4.orig	2014-03-12 18:24:31.013871791 +0100
+++ sssd-1.11.4/src/external/python.m4	2014-03-12 19:55:39.160308983 +0100
@@ -12,15 +12,8 @@
     AC_PATH_PROG(PYTHON, python)
     AC_MSG_CHECKING([for working python])
     if test -x "$PYTHON"; then
-        PYTHON_CFLAGS="`$PYTHON -c \"from distutils import sysconfig; \
-            print '-I' + sysconfig.get_python_inc() + \
-            ' -I' + sysconfig.get_python_inc(plat_specific=True) + ' ' + \
-            sysconfig.get_config_var('BASECFLAGS')\"`"
-        PYTHON_LIBS="`$PYTHON -c \"from distutils import sysconfig; \
-            print \\\" \\\".join(sysconfig.get_config_var('LIBS').split() + \
-            sysconfig.get_config_var('SYSLIBS').split()) + \
-            ' -lpython' + sysconfig.get_config_var('VERSION') + \
-            ' -L' + sysconfig.get_config_var('LIBDIR')\"`"
+        PYTHON_CFLAGS="`python-config --cflags`"
+        PYTHON_LIBS="`python-config --libs`"
             AC_MSG_RESULT([yes])
     else
         AC_MSG_RESULT([no])
