--- sssd-1.12.3/Makefile.am.orig	2015-01-12 18:49:06.135356150 +0100
+++ sssd-1.12.3/Makefile.am	2015-01-12 19:16:19.545287606 +0100
@@ -870,7 +870,7 @@
     src/sss_client/libwbclient/wbc_util_sssd.c
 libwbclient_la_LIBADD = \
     libsss_nss_idmap.la \
-    $(CLIENT_LIBS)
+    $(CLIENT_LIBS) -ldl
 libwbclient_la_LDFLAGS = \
     -Wl,--version-script,$(srcdir)/src/sss_client/libwbclient/wbclient.exports \
     -version-info 11:0:11
