diff -urNp -x '*.orig' sssd-1.13.4.org/Makefile.am sssd-1.13.4/Makefile.am
--- sssd-1.13.4.org/Makefile.am	2021-03-03 21:59:13.565804202 +0100
+++ sssd-1.13.4/Makefile.am	2021-03-03 21:59:13.715851718 +0100
@@ -3455,7 +3455,8 @@ if WITH_JOURNALD
     systemdconf_DATA += \
         src/sysv/systemd/journal.conf
 endif
-else
+endif
+if HAVE_SYSV
 if HAVE_SUSE
     init_SCRIPTS += \
         src/sysv/SUSE/sssd
diff -urNp -x '*.orig' sssd-1.13.4.org/configure.ac sssd-1.13.4/configure.ac
--- sssd-1.13.4.org/configure.ac	2016-04-13 16:48:41.000000000 +0200
+++ sssd-1.13.4/configure.ac	2021-03-03 21:59:13.715851718 +0100
@@ -207,7 +207,7 @@ if test x$HAVE_NSCD; then
 fi
 
 WITH_INITSCRIPT
-if test x$initscript = xsystemd; then
+if test x"${initscript%systemd}" != x"${initscript}"; then
     WITH_SYSTEMD_UNIT_DIR
     WITH_SYSTEMD_CONF_DIR
 fi
diff -urNp -x '*.orig' sssd-1.13.4.org/src/conf_macros.m4 sssd-1.13.4/src/conf_macros.m4
--- sssd-1.13.4.org/src/conf_macros.m4	2016-04-13 16:48:41.000000000 +0200
+++ sssd-1.13.4/src/conf_macros.m4	2021-03-03 21:59:13.715851718 +0100
@@ -130,14 +130,15 @@ AC_DEFUN([WITH_INITSCRIPT],
   fi
 
   if test x"$with_initscript" = xsysv || \
-     test x"$with_initscript" = xsystemd; then
+     test x"$with_initscript" = xsystemd || \
+     test x"$with_initscript" = xsysv,systemd; then
         initscript=$with_initscript
   else
       AC_MSG_ERROR([Illegal value -$with_initscript- for option --with-initscript])
   fi
 
-  AM_CONDITIONAL([HAVE_SYSV], [test x"$initscript" = xsysv])
-  AM_CONDITIONAL([HAVE_SYSTEMD_UNIT], [test x"$initscript" = xsystemd])
+  AM_CONDITIONAL([HAVE_SYSV], [test x"${initscript#sysv}" != "x${initscript}"])
+  AM_CONDITIONAL([HAVE_SYSTEMD_UNIT], [test x"${initscript%systemd}" != x"${initscript}"])
   AC_MSG_NOTICE([Will use init script type: $initscript])
   ])
 
